### PR TITLE
kafka functional test specific changes

### PR DESCRIPTION
**Pull request by pmoogi-redhat**
###Description
Moving kafka output testing to the functional framework,
so that it can speed up e2e testing and isolate kafka output testing from uncontrollable cluster variables.

Scope of this functional test - validating if we can produce a fluentd configuration that allows fluent to write messages to kafka. This only requires that fluent be able to: 1. connect to kafka 2. push a message

/cc @jcantrill @vimalk78
/assign @jcantrill

Links
Depending on PR(s):
Bugzilla:
Github issue:
JIRA: https://issues.redhat.com/browse/LOG-1587

### CODE REVIEW

File: test/runtime/pod.go [hunk 1]
```diff
@@ -125,3 +186,50 @@ func (builder *PodBuilder) AddLabels(labels map[string]string) *PodBuilder {
 	}
 	return builder
 }
+
+func (builder *ContainerBuilder) AddContainerPort(name string, port int32) *ContainerBuilder {
+	builder.container.Ports = append(builder.container.Ports, corev1.ContainerPort{Name: name, ContainerPort: port})
+	return builder
+}
+
+func (builder *ContainerBuilder) AddEnvVarFromEnvVarSourceNodeToInitContainer(name string, initContainerno int) *ContainerBuilder {
```
Comment (vimalk78) [inline test/runtime/pod.go:195]: long function name. seems to be doing very specific thing. can you break into smaller reusable functions and compose them together?
perhaps smaller function can be reused in `AddEnvVarFromEnvVarSourcePodToInitContainer` also

File: test/runtime/pod.go [hunk 2]
```diff
@@ -125,3 +186,50 @@ func (builder *PodBuilder) AddLabels(labels map[string]string) *PodBuilder {
...
+			FieldRef: &corev1.ObjectFieldSelector{
+				FieldPath: "spec.nodeName",
+			},
+		},
+	})
+
+	return builder
+}
+
+func (builder *ContainerBuilder) AddEnvVarFromEnvVarSourcePodToInitContainer(name string, initContainerno int) *ContainerBuilder {
```
Comment (vimalk78) [inline test/runtime/pod.go:209]: long, very specific function name. follow the comment above and break

File: test/runtime/pod.go [hunk 3]
```diff
@@ -125,3 +186,50 @@ func (builder *PodBuilder) AddLabels(labels map[string]string) *PodBuilder {
...
+			FieldRef: &corev1.ObjectFieldSelector{
+				FieldPath: "metadata.name",
+			},
+		},
+	})
+
+	return builder
+}
+
+func (builder *ContainerBuilder) AddEnvVarFromEnvVarSourceNamespaceToInitContainer(name string, initContainerno int) *ContainerBuilder {
```
Comment (vimalk78) [inline test/runtime/pod.go:223]: same as above

File: test/runtime/pod.go [hunk 4]
```diff
@@ -107,6 +146,28 @@ func (builder *PodBuilder) AddConfigMapVolume(name, configMapName string) *PodBu
 	return builder
 }
 
+func (builder *PodBuilder) AddEmptyDirVolume(name string) *PodBuilder {
```
Comment (vimalk78) [inline test/runtime/pod.go:149]: suggest to refactor the three functions into one. `AddVolumeFromSource(name, VolumeSource)`

File: test/functional/outputs/forward_to_kafka_test.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		f2 func(spec *logging.OutputSpec)) func(*logging.OutputSpec) {
+		return func(s *logging.OutputSpec) {
+			f1(s)
+			f2(s)
+		}
+	}
+
+	//	timestamp := "2013-03-28T14:36:03.243000+00:00"
+
+	Context("Application Logs", func() {
```
[the remark this answers is not in the document]
Comment (jcantrill) [reply test/functional/outputs/forward_to_kafka_test.go:43]: ... we deleted the all logs to single topic test.... at least trying to verify they all got there. I we probably dont need that case

File: test/functional/outputs/forward_to_kafka_test.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		FIt("should send large message over Kafka", func() {
+			functional.NewClusterLogForwarderBuilder(framework.Forwarder).
+				FromInput(logging.InputNameApplication).
+				ToOutputWithVisitor(join(setKafkaSpecValues, func(spec *logging.OutputSpec) {
```
Comment (jcantrill) [inline test/functional/outputs/forward_to_kafka_test.go:47]: I thought we wrote a function "ToKafka" that would cover this

File: test/functional/outputs/forward_to_kafka_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (eranra) [reply test/functional/outputs/forward_to_kafka_test.go:43]: @jcantrill @vimalk78 I talked with @pmoogi-redhat on adding to this PR the process of removing e2e Kafka test. Please advice. From your above remarks (and logic) it makes sense that this "replaces" the e2e tests for Kafka hence should remove them when this merges

File: test/functional/outputs/forward_to_kafka_test.go (hunk 6 above)
Comment (eranra) [reply test/functional/outputs/forward_to_kafka_test.go:47]: And when you move this to "ToKafka" you will be able to delete also the join and setKafkaSpecValues functions

File: test/functional/outputs/forward_to_kafka_test.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+					//at this port fluent connects with kafka broker
+					spec.URL = "https://localhost:9093"
+				}), logging.OutputTypeKafka)
+			Expect(framework.Deploy()).To(BeNil())
+
+			var MaxLen uint64 = 40000
```
Comment (eranra) [inline test/functional/outputs/forward_to_kafka_test.go:53]: I think that the value 40,000 had some meaning for syslog ... not sure there is a need for that in this firstKafaka test. you can use "WritesApplicationLogs" instead (this writes fix size of 100 bytes)

File: test/functional/framework.go [hunk 8]
```diff
@@ -470,6 +486,27 @@ func (f *FluentdFunctionalFramework) ReadApplicationLogsFrom(outputName string)
 	return f.ReadLogsFrom(outputName, applicationLog)
 }
 
+func (f *FluentdFunctionalFramework) ReadApplicationLogsFromKafka(topic string, brokerlistener string, consumercontainername string) (results []string, err error) {
+	//inter broker zookeeper connect is plaintext so use plaintext port to check on sent messages from kafka producer ie. fluent-kafka plugin
```
Comment (eranra) [inline test/functional/framework.go:490]: I see that you are reading from a file "/shared/consumed.logs" in this function. If this is the case, you can use the function "ReadLogsFrom" and you do not need to reimplement it. You need to add the needed values to outputLogFile and than this function will work like all the rest of the wrappers.

File: test/functional/output_kafka.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	// Deploy Zookeeper steps : create configmap, create container, create zookeeper service
+
+	//zookeeperm - configmap with zookeeperDeploymentName being "zookeeper", in b.Pod.Namespace
+	zookeepercm := kafka.NewZookeeperConfigMapFunctionalTestPod(b.Pod.Namespace)
+	log.V(2).Info("Creating zookeeper configmap", "namespace", zookeepercm.Namespace, "name", zookeepercm.Name)
+	if err := f.Test.Client.Create(zookeepercm); err != nil {
+		return err
+	}
+
+	//init container for zookeep
```
Comment (eranra) [inline test/functional/output_kafka.go:48]: s/zookeep/zookeeper

File: test/functional/output_kafka.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+		AddContainerPort("client", zookeeperClientPort).
+		AddContainerPort("peer", zookeeperPeerPort).
+		AddContainerPort("leader-election", zookeeperLeaderElectionPort).
+		AddVolumeMount("configmapkafka", "/etc/kafka-configmap", "", false).
+		AddVolumeMount("configkafka", "/etc/kafka", "", false).
+		AddVolumeMount("datazookeeper", "/var/lib/zookeeper", "", false).
+		WithCmd("./bin/zookeeper-server-start.sh /etc/kafka/zookeeper.properties").
+		End().
+		AddConfigMapVolume("configmapkafka", zookeeperDeploymentName).
+		AddEmptyDirVolume("configkafka").
```
Comment (eranra) [inline test/functional/output_kafka.go:68]: I am a bit confused with the volumes and names ... we have two volumes added here "configmapkafka" and "configkafka" one has the zookeeperDeploymentName and the second one is empty ... can you update the names of the volumes so they will be relevant to the content.

File: test/functional/output_kafka.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+		AddEmptyDirVolume("zookeeperlogs").
+		AddEmptyDirVolume("datazookeeper")
```
Comment (eranra) [inline test/functional/output_kafka.go:70]: I suggest s/"zookeeperlogs"/"zookeeper-logs" and s/"datazookeeper/"zookeeper-data" etc ... so the volume name is prefixed with app ... another option is to drop the "zookeeper" prefixed from the names altogether. As it is now it is confusing.

File: test/functional/output_kafka.go [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		AddEnvVar("extensions", "/opt/kafka/libs/extensions").
+		AddEnvVar("data", "/var/lib/kafka/data").
+		End().
+		AddConfigMapVolume("brokerconfig", kafka.DeploymentName).
+		AddSecretVolume("brokercerts", kafka.DeploymentName).
+		AddEmptyDirVolume("brokerlogs").
+		AddEmptyDirVolume("extensions")
+
+	/////////////////////////////////////////////
+	//step c
```
Comment (eranra) [inline test/functional/output_kafka.go:125]: missing the remark that explains what you do in this step

File: test/functional/output_kafka.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+			AddVolumeMount("brokerconfig", "/etc/kafka-configmap", "", false).
+			AddVolumeMount("brokercert", "/etc/kafka-certs", "", false).
+			AddVolumeMount("shared", "/shared", "", false).
+			End().
+			AddSecretVolume("brokercert", kafka.DeploymentName).
+			AddEmptyDirVolume("shared")
+
+	}
+
+	return nil
```
Comment (eranra) [inline test/functional/output_kafka.go:152]: I propose to break this function into 3 smaller functions and not have one big function for all steps.

File: test/functional/output_kafka.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+	for _, topic := range topics {
+		containername := kafka.ConsumerNameForTopic(topic)
+		// Deploy consumer
+		//consumerapp := kafka.NewKafkaConsumerDeployment(OpenshiftLoggingNS, topic)
+
+		log.V(2).Info("Creating Broker Consumer app")
+		///create topic using --bootstrap-server localhost:9093
+		cmdCreateTopic := fmt.Sprintf(`./bin/kafka-topics.sh --zookeeper localhost:2181 --create --if-not-exists --topic %s --partitions 1 --replication-factor 1 ;`, topic)
+		cmdRunConsumer := fmt.Sprintf(`./bin/kafka-console-consumer.sh --bootstrap-server %s --topic %s --from-beginning > /shared/consumed.logs ;`, "localhost:9092", topic)
+		cmdSlice := []string{"sleep 120;", cmdCreateTopic, cmdRunConsumer}
```
Comment (eranra) [inline test/functional/output_kafka.go:137]: Do we really need to wait two minutes ... this adds a lot of time to the test ... can this be polling/wait for something stage somehow and not sleep. The idea of functional tests are to be quick ... so 120 seconds might be a lot of time here.

File: internal/generator/fluentd/output/kafka/kafka.go [hunk 15]
```diff
@@ -41,6 +41,10 @@ func (k Kafka) Template() string {
 brokers {{.Brokers}}
 default_topic {{.Topics}}
 use_event_time true
+ssl_client_cert_key '/etc/kafka-certs/tls.key'
+ssl_client_cert '/etc/kafka-certs/tls.crt'
+ssl_ca_cert '/etc/kafka-certs/ca-bundle.crt'
+sasl_over_ssl false
```
Comment (vimalk78) [inline internal/generator/fluentd/output/kafka/kafka.go:47]: these keys are already supported in `SecurityConfig` field

File: internal/runtime/pod.go [hunk 16]
```diff
@@ -151,3 +157,98 @@ func (builder *PodBuilder) AddLabels(labels map[string]string) *PodBuilder {
...
+	return builder
+}
+
+func (builder *ContainerBuilder) WithCmdStringSlice(cmdAgrgs []string) *ContainerBuilder {
+	builder.container.Command = []string{cmdAgrgs[0]}
+	builder.container.Args = cmdAgrgs[1:]
+	return builder
+}
+
+func (builder *ContainerBuilder) AddEnvVarToInitContainer(name, value string, initContainerno int) *ContainerBuilder {
```
Comment (jcantrill) [inline internal/runtime/pod.go:220]: InitContainerBuilder#AddEnvVar

File: internal/runtime/pod.go [hunk 17]
```diff
@@ -151,3 +157,98 @@ func (builder *PodBuilder) AddLabels(labels map[string]string) *PodBuilder {
 	}
 	return builder
 }
+
+func (builder *ContainerBuilder) AddEnvVarFromEnvVarSourceNodeToInitContainer(name string, initContainerno int) *ContainerBuilder {
```
Comment (jcantrill) [inline internal/runtime/pod.go:161]: InitContainerBuilder

File: test/framework/functional/cluster_log_forwarder.go [hunk 18]
```diff
@@ -92,9 +96,19 @@ func (p *PipelineBuilder) ToOutputWithVisitor(visit OutputSpecVisiter, outputNam
 					Name: "cloudwatch",
 				},
 			}
+		case logging.OutputTypeKafka:
+			output = &logging.OutputSpec{
+				Name: logging.OutputTypeKafka,
+				Type: logging.OutputTypeKafka,
+				URL:  "https://0.0.0.0:9093",
+				Secret: &logging.OutputSecretSpec{
```
Comment (jcantrill) [inline test/framework/functional/cluster_log_forwarder.go:104]: Reminde me?  The only way you were able to make this work was with a secret?

File: test/framework/functional/framework.go [hunk 19]
```diff
@@ -210,6 +217,9 @@ func (f *FluentdFunctionalFramework) DeployWithVisitors(visitors []runtime.PodBu
 	if f.Conf, err = forwarder.Generate(string(clfYaml), false, debugOutput, &testClient); err != nil {
 		return err
 	}
+	//overriding fluent config with Kafka specifics tls.key, crts, ca settings, this needs to done at Generator level though
```
Comment (jcantrill) [inline test/framework/functional/framework.go:220]: remove me if not needed

File: test/functional/outputs/forward_to_kafka_test.go (hunk 6 above)
Comment (jcantrill) [reply test/functional/outputs/forward_to_kafka_test.go:54]: I think this comment is still valid

File: test/helpers/kafka/broker.go [hunk 20]
```diff
@@ -356,3 +366,26 @@ func NewBrokerSecret(namespace string) *v1.Secret {
 	)
 	return secret
 }
+
+func NewBrokerSecretFunctionalTestPod(namespace string) *v1.Secret {
```
Comment (jcantrill) [inline test/helpers/kafka/broker.go:370]: Shouldnt this just Be "NewBrokerSecret"

**Comment by jcantrill** [comment]
closing i favor of https://github.com/openshift/cluster-logging-operator/pull/1237

**Pull request closed without merging**